diff -u -r unbound-1.5.1-orig/libunbound/context.c unbound-1.5.1/libunbound/context.c
--- unbound-1.5.1-orig/libunbound/context.c	2014-02-07 14:28:39.000000000 +0100
+++ unbound-1.5.1/libunbound/context.c	2015-02-09 11:55:02.000000000 +0100
@@ -125,7 +125,7 @@
 }
 
 struct ctx_query* 
-context_new(struct ub_ctx* ctx, const char* name, int rrtype, int rrclass, 
+context_create(struct ub_ctx* ctx, const char* name, int rrtype, int rrclass, 
 	ub_callback_t cb, void* cbarg)
 {
 	struct ctx_query* q = (struct ctx_query*)calloc(1, sizeof(*q));
diff -u -r unbound-1.5.1-orig/libunbound/context.h unbound-1.5.1/libunbound/context.h
--- unbound-1.5.1-orig/libunbound/context.h	2014-02-07 14:28:39.000000000 +0100
+++ unbound-1.5.1/libunbound/context.h	2015-02-09 11:54:44.000000000 +0100
@@ -241,7 +241,7 @@
  * @param cbarg: user arg for async queries.
  * @return new ctx_query or NULL for malloc failure.
  */
-struct ctx_query* context_new(struct ub_ctx* ctx, const char* name, int rrtype,
+struct ctx_query* context_create(struct ub_ctx* ctx, const char* name, int rrtype,
         int rrclass, ub_callback_t cb, void* cbarg);
 
 /**
diff -u -r unbound-1.5.1-orig/libunbound/libunbound.c unbound-1.5.1/libunbound/libunbound.c
--- unbound-1.5.1-orig/libunbound/libunbound.c	2014-11-04 10:11:59.000000000 +0100
+++ unbound-1.5.1/libunbound/libunbound.c	2015-02-09 11:55:20.000000000 +0100
@@ -634,7 +634,7 @@
 	}
 	/* create new ctx_query and attempt to add to the list */
 	lock_basic_unlock(&ctx->cfglock);
-	q = context_new(ctx, name, rrtype, rrclass, NULL, NULL);
+	q = context_create(ctx, name, rrtype, rrclass, NULL, NULL);
 	if(!q)
 		return UB_NOMEM;
 	/* become a resolver thread for a bit */
@@ -687,7 +687,7 @@
 	}
 
 	/* create new ctx_query and attempt to add to the list */
-	q = context_new(ctx, name, rrtype, rrclass, (ub_callback_t)callback,
+	q = context_create(ctx, name, rrtype, rrclass, (ub_callback_t)callback,
 		mydata);
 	if(!q)
 		return UB_NOMEM;
@@ -733,7 +733,7 @@
 	}
 
 	/* create new ctx_query and attempt to add to the list */
-	q = context_new(ctx, name, rrtype, rrclass, callback, mydata);
+	q = context_create(ctx, name, rrtype, rrclass, callback, mydata);
 	if(!q)
 		return UB_NOMEM;
 
