Index: libavcodec/libx264.c
===================================================================
--- libavcodec/libx264.c	(révision 20783)
+++ libavcodec/libx264.c	(copie de travail)
@@ -99,7 +99,7 @@
         }
 
         x4->pic.i_pts  = frame->pts;
-        x4->pic.i_type = X264_TYPE_AUTO;
+        x4->pic.i_type = frame->key_frame ? X264_TYPE_IDR : X264_TYPE_AUTO;
     }
 
     if (x264_encoder_encode(x4->enc, &nal, &nnal, frame? &x4->pic: NULL, &pic_out) < 0)
