Index: libswscale/swscale_template.c
===================================================================
--- libswscale/swscale_template.c (révision 31187)
+++ libswscale/swscale_template.c (copie de travail)
@@ -2252,7 +2252,7 @@
     "shrl       $9, %%esi    \n\t"                                              \
 
 static inline void RENAME(hyscale_fast)(SwsContext *c, int16_t *dst,
-                                        long dstWidth, const uint8_t *src, int srcW,
+                                        x86_reg dstWidth, const uint8_t *src, int srcW,
                                         int xInc)
 {
 #if ARCH_X86
@@ -2401,7 +2401,7 @@
 }
 
 static inline void RENAME(hcscale_fast)(SwsContext *c, int16_t *dst,
-                                        long dstWidth, const uint8_t *src1,
+                                        x86_reg dstWidth, const uint8_t *src1,
                                         const uint8_t *src2, int srcW, int xInc)
 {
 #if ARCH_X86
