diff -ru a/x264-snapshot-20091208-2245/common/cabac.c b/x264-snapshot-20091208-2245/common/cabac.c
--- a/x264-snapshot-20091208-2245/common/cabac.c	2009-12-08 22:45:08.000000000 +0100
+++ b/x264-snapshot-20091208-2245/common/cabac.c	2010-01-08 09:11:43.000000000 +0100
@@ -666,7 +666,7 @@
 
 /* FIXME could avoid this duplication by reversing the order of states
  * with MPS=0, but that would uglify the other tables */
-const uint8_t x264_cabac_range_lps[128][4] =
+const uint8_t x264_cabac_range_lps[128][4] __attribute__((visibility("hidden"))) = 
 {
     {   2,   2,   2,   2 },
     {   6,   7,   8,   9 }, {   6,   7,   9,  10 }, {   6,   8,   9,  11 },
@@ -715,7 +715,7 @@
     {   2,   2,   2,   2 },
 };
 
-const uint8_t x264_cabac_transition[128][2] =
+const uint8_t x264_cabac_transition[128][2] __attribute__((visibility("hidden"))) = 
 {
     {  0,  0}, {  1, 25}, {  1, 25}, {  2, 26}, {  3, 26}, {  4, 26}, {  5, 27}, {  6, 27},
     {  7, 27}, {  8, 28}, {  9, 28}, { 10, 28}, { 11, 29}, { 12, 29}, { 13, 30}, { 14, 30},
@@ -735,7 +735,7 @@
     {100,121}, {100,122}, {101,123}, {101,124}, {101,125}, {102,126}, {102,126}, {127,127},
 };
 
-const uint8_t x264_cabac_renorm_shift[64]= {
+const uint8_t x264_cabac_renorm_shift[64]  __attribute__((visibility("hidden"))) = {
  6,5,4,4,3,3,3,3,2,2,2,2,2,2,2,2,
  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
diff -ru a/x264-snapshot-20091208-2245/common/quant.c b/x264-snapshot-20091208-2245/common/quant.c
--- a/x264-snapshot-20091208-2245/common/quant.c	2009-12-08 22:45:08.000000000 +0100
+++ b/x264-snapshot-20091208-2245/common/quant.c	2010-01-08 09:34:09.000000000 +0100
@@ -163,9 +163,9 @@
  *  chroma: for the complete mb: if score < 7 -> null
  */
 
-const uint8_t x264_decimate_table4[16] = {
+const uint8_t x264_decimate_table4[16]  __attribute__((visibility("hidden"))) = {
     3,2,2,1,1,1,0,0,0,0,0,0,0,0,0,0 };
-const uint8_t x264_decimate_table8[64] = {
+const uint8_t x264_decimate_table8[64]  __attribute__((visibility("hidden"))) = {
     3,3,3,3,2,2,2,2,2,2,2,2,1,1,1,1,
     1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,
     0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
diff -ru a/x264-snapshot-20091208-2245/common/x86/dct-32.asm b/x264-snapshot-20091208-2245/common/x86/dct-32.asm
--- a/x264-snapshot-20091208-2245/common/x86/dct-32.asm	2009-12-08 22:45:08.000000000 +0100
+++ b/x264-snapshot-20091208-2245/common/x86/dct-32.asm	2010-01-08 11:23:31.000000000 +0100
@@ -191,7 +191,7 @@
 ; void x264_sub8x8_dct8_mmx( int16_t dct[8][8], uint8_t *pix1, uint8_t *pix2 )
 ;-----------------------------------------------------------------------------
 cglobal x264_sub8x8_dct8_mmx, 3,3
-global x264_sub8x8_dct8_mmx.skip_prologue
+global x264_sub8x8_dct8_mmx.skip_prologue:hidden
 .skip_prologue:
     INIT_MMX
     call load_diff_4x8_mmx
@@ -257,7 +257,7 @@
 ; void x264_add8x8_idct8_mmx( uint8_t *dst, int16_t dct[8][8] )
 ;-----------------------------------------------------------------------------
 cglobal x264_add8x8_idct8_mmx, 2,2
-global x264_add8x8_idct8_mmx.skip_prologue
+global x264_add8x8_idct8_mmx.skip_prologue:hidden
 .skip_prologue:
     INIT_MMX
     add word [r1], 32
@@ -346,7 +346,7 @@
 %macro DCT_SUB8 1
 cglobal x264_sub8x8_dct_%1, 3,3
     add r2, 4*FDEC_STRIDE
-global x264_sub8x8_dct_%1.skip_prologue
+global x264_sub8x8_dct_%1.skip_prologue:hidden
 .skip_prologue:
 %ifnidn %1, sse2
     mova m7, [hsub_mul GLOBAL]
@@ -379,7 +379,7 @@
 ;-----------------------------------------------------------------------------
 cglobal x264_sub8x8_dct8_%1, 3,3
     add r2, 4*FDEC_STRIDE
-global x264_sub8x8_dct8_%1.skip_prologue
+global x264_sub8x8_dct8_%1.skip_prologue:hidden
 .skip_prologue:
 %ifidn %1, sse2
     LOAD_DIFF m0, m7, none, [r1+0*FENC_STRIDE], [r2-4*FDEC_STRIDE]
@@ -423,7 +423,7 @@
 ;-----------------------------------------------------------------------------
 cglobal x264_add8x8_idct_sse2, 2,2
     add r0, 4*FDEC_STRIDE
-global x264_add8x8_idct_sse2.skip_prologue
+global x264_add8x8_idct_sse2.skip_prologue:hidden
 .skip_prologue:
     UNSPILL_SHUFFLE r1, 0,2,1,3, 0,1,2,3
     SBUTTERFLY qdq, 0, 1, 4
@@ -460,7 +460,7 @@
 ;-----------------------------------------------------------------------------
 cglobal x264_add8x8_idct8_sse2, 2,2
     add r0, 4*FDEC_STRIDE
-global x264_add8x8_idct8_sse2.skip_prologue
+global x264_add8x8_idct8_sse2.skip_prologue:hidden
 .skip_prologue:
     UNSPILL r1, 1,2,3,5,6,7
     IDCT8_1D   0,1,2,3,4,5,6,7,r1
diff -ru a/x264-snapshot-20091208-2245/common/x86/dct-64.asm b/x264-snapshot-20091208-2245/common/x86/dct-64.asm
--- a/x264-snapshot-20091208-2245/common/x86/dct-64.asm	2009-12-08 22:45:08.000000000 +0100
+++ b/x264-snapshot-20091208-2245/common/x86/dct-64.asm	2010-01-08 09:32:27.000000000 +0100
@@ -149,7 +149,7 @@
     call .skip_prologue
     RET
 %endif
-global x264_sub8x8_dct_%1.skip_prologue
+global x264_sub8x8_dct_%1.skip_prologue:hidden
 .skip_prologue:
     SWAP 7, 9
     LOAD_DIFF8x4 0, 1, 2, 3, 8, 9, r1, r2-4*FDEC_STRIDE
@@ -176,7 +176,7 @@
     call .skip_prologue
     RET
 %endif
-global x264_sub8x8_dct8_%1.skip_prologue
+global x264_sub8x8_dct8_%1.skip_prologue:hidden
 .skip_prologue:
     SWAP 7, 10
     LOAD_DIFF8x4  0, 1, 2, 3, 4, 10, r1, r2-4*FDEC_STRIDE
@@ -214,7 +214,7 @@
     call .skip_prologue
     RET
 %endif
-global x264_add8x8_idct8_sse2.skip_prologue
+global x264_add8x8_idct8_sse2.skip_prologue:hidden
 .skip_prologue:
     SWAP 7, 9
     movdqa  m0, [r1+0x00]
@@ -246,7 +246,7 @@
     call .skip_prologue
     RET
 %endif
-global x264_add8x8_idct_sse2.skip_prologue
+global x264_add8x8_idct_sse2.skip_prologue:hidden
 .skip_prologue:
     SWAP 7, 9
     mova   m0, [r1+ 0]
